libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.h

```diff
@@ -0,0 +1,201 @@
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+    // Return AP_PacketDigitalCAN from @driver_index or nullptr if it's not ready or doesn't exist
+    static AP_PacketDigitalCAN *get_pdcan(uint8_t driver_index);
+
+    // initialize AP_PacketDigitalCAN bus
+    void init(uint8_t driver_index, bool enable_filters) override;
+    bool add_interface(AP_HAL::CANIface* can_iface) override;
+
+    // test if the AP_PacketDigital CAN driver is ready to be armed
+    bool pre_arm_check(char* reason, uint8_t reason_len) const;
+
+    bool get_voltage_and_current_and_temp(const uint16_t serialnumber, const bool chkInput, float &voltage, float &current, float &temperature) const;
+
+    bool is_healthy(const uint16_t serialnumber) const {
+        return is_healthy_by_index(get_device_index(serialnumber));
+    }
+
+private:
+
+    // Message groups form part of the CAN ID of each frame
+    enum class PacketType : uint8_t {
+        UNKNOWN         = 0,
+        ACK             = 0x21,
+        NACK            = 0x22,
+        PING            = 0x23,
+        FAULT           = 0x42,
+        ALGORITHM_SET   = 0x44,
+        ALGORITHM_GET   = 0x45,
+        STREAM_FAULT    = 0x47,
+        STREAM_INPUT    = 0x48,
+        STREAM_OUTPUT   = 0x49,
+        VOLTAGE_SET     = 0x4B,
+        VOLTAGE_GET     = 0x4C,
+        CVT_SET         = 0x4D,
+        CVT_GET         = 0x4E,
+    };
+
+    enum class Fault_Flags : uint8_t {
+        NONE                = 0,
+        OVER_VOLTAGE        = (1<<0),
+        UNDER_VOLTAGE       = (1<<1),
+        OVER_CURRENT        = (1<<2),
+        OVER_TEMPERATURE    = (1<<3),
+    };
+
+    enum class Config_Mode : int8_t {
+        DISABLED            = 0,
+        REPORT_INPUTS       = 1,
+        REPORT_OUTPUTS      = 2,
+        REPORT_VOUT_CVT_ALG = 3,
+        SET_VOUT            = 4,
+        SET_CVT             = 5,
+        SET_ALGORITHM       = 6,
+    };
+
+    // loop to send output to ESCs in background thread
+    void loop();
+
+    // write frame on CAN bus, returns true on success
+    bool write_frame(AP_HAL::CANFrame &out_frame, const uint64_t timeout);
+
+    // read frame on CAN bus, returns true on success
+    bool read_frame(AP_HAL::CANFrame &recv_frame, const uint64_t timeout, AP_HAL::CANIface::CanIOFlags &flags);
+
+    // parse inbound frames
+    void handle_frame(const AP_HAL::CANFrame &frame);
+
+    // find first MPPT_devices[] index that contains this serialnumber. If not found, returns UINT8_MAX
+    uint8_t get_device_index(const uint16_t serial_number) const;
+
+    // send command to MPPT device
+    void send_command(const PacketType type, const uint16_t serialnumber, const float data = 0.0f);
+
+    bool is_healthy_by_index(const uint8_t index) const {
+        if (index >= device_count) {
+            return false;
+        }
+//        if (MPPT_devices[index].faults != Fault_Flags::NONE) {
+//            // TODO: test this
+//            return false;
+//        }
```

Comment:
Are you planning on doing this before merge?

Reply:
removed